[MacrosOnImports][Swiftify] Copy module imports from clang node's module to its Swift macro SourceFile

hnrklssn opened the pull request:
Since the _SwiftifyImport peer macro expansion reuses the function signature of the original imported function, it needs to import the same modules as the module where the original clang node resides, to ensure visibility of any symbols that may appear in the signature.

rdar://151611573

hnrklssn committed:
[ClangImporter] Add clang module imports to Swift wrapper module

When macros like _SwiftifyImport are added to a wrapper module for a
clang module, they may need to refer to symbols declared in another
clang module that the wrapped module imports (e.g. because they are used
in the original signature). This adds all the imported clang modules as
implicit imports to the wrapper module.

rdar://151611573

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented on test/Interop/C/swiftify-import/clang-includes-noexport.swift:
```diff
@@ -0,0 +1,60 @@
+
+// CHECK:      import ModuleA
+// CHECK-NEXT: import ModuleB
+// CHECK-NOT:  import
```
@DougGregor I don't fully get why only the module's imports that themselves are top-level modules show up in the dump. It seems to work regardless, but it seems a bit off. Do you know if this is intended?

hnrklssn replied:
Somewhat fixed by https://github.com/swiftlang/swift/pull/81859/commits/a56810277defa28d8c950bec1915359ecd56ecd6
@bnbarham I think this affects IDE stuff, so I didn't dare make it actually print every module even though it would've been kinda nice if this reflected the full imports. Do you have more context for why this is the way it is?

hnrklssn committed:
fix implicit import of CxxStd module

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test windows

hnrklssn committed:
fix dangling StringRef

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented on test/Interop/C/swiftify-import/Inputs/clang-includes.h:
```diff
@@ -0,0 +1,20 @@
+#define __counted_by(x) __attribute__((__counted_by__(x)))
+#define __noescape __attribute__((noescape))
+
+void basic_include(const a_t *__counted_by(len) p __noescape, a_t len);
```
The compiler only strips away typedefs that surround the trainsformed pointer type. Pointee typedefs remain.

fahadnayyar commented:
@swift-ci please smoke test macOS Linux

hnrklssn commented:
@swift-ci please smoke test macos

hnrklssn commented:
@swift-ci please smoke test macos

hnrklssn commented:
@swift-ci please smoke test macos

fahadnayyar commented:
@swift-ci please smoke test

hnrklssn commented:
https://github.com/swiftlang/llvm-project/pull/10957

@swift-ci please smoke test

hnrklssn commented:
https://github.com/swiftlang/llvm-project/pull/10957

@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
fix tests after rebase

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
https://github.com/swiftlang/llvm-project/pull/10957
@swift-ci please smoke test

hnrklssn commented:
https://github.com/swiftlang/llvm-project/pull/11022
@swift-ci please smoke test

hnrklssn committed:
[ClangImporter] Add submodules as implicit imports of wrapper module

Decls in Swift wrapper module may not originate in the top-level clang
module with the same name, since decls in clang submodules are dumped
into the top-level module. This is because Swift has no concept of
submodules. To make sure that any imported decl has access to the same
symbols as the original clang decl had, all transitive submodules, and
their imports, are added as implicit imports of the wrapper module. This
is necessary in the case where a submodule is marked `explicit`.

The content in explicit submodules isn't normally made visible when
importing the parent module. Decls from explicit submodules still
end up in the top-level wrapper module however, so in Swift they do
still need to be visible from the top-level module. This is relevant
for macro expansions, so that they can refer to the same types as the
original decl.

hnrklssn committed:
test cxx mode in clang-includes tests

hnrklssn commented:
https://github.com/swiftlang/llvm-project/pull/11022
@swift-ci please smoke test

DougGregor reviewed (approved):
This direction looks good, but please implement the two small optimizations I suggest.

hnrklssn committed:
Add hidden --dump-source-file-imports flag

This flag dumps all imports for each SourceFile after it's gone through
import resolution. It is only intended for testing purposes.
There are other ways to print imports, but they don't correspond 1:1 to
the imports actually resolved, which is a bit problematic when testing
implicit clang module imports.

hnrklssn committed:
[ClangImporter] Only import explicit submodules to the current module

Non-explicit submodules don't need to be explicitly added to the list of
imports to be visible, since their decls are implicitly exported. Skip
these modules even when present in the list of imports. Explicit
submodules are imported *regardless* of whether another module
imports them however.

hnrklssn committed:
[ImportResolution] Deduplicate top-level clang modules from import list

Importing clang submodules results in an implicit import of the
top-level module as well. This can result in the same TLM being imported
many different times, if multiple submodules are imported from the same
module. This deduplicates these imports.

Other imports are not expected to be duplicated, so care is taken to
only deduplicate clang TLM imports.

hnrklssn committed:
[ImportResolution] Remove redundant "Swift" module import

Import resolution now adds the Swift module to the list of imports, so
we no longer need to add it manually.

hnrklssn committed:
[ImportResolution] Remove early exit that will never trigger

There's only one caller to performImportResolutionForClangMacroBuffer,
which always passes a newly allocated SourceFile. This removes a
redundant check for whether its imports have been resolved, and instead
asserts that they haven't.

hnrklssn committed:
[SwiftifyImport] Add various submodule import test cases

Updates to these test cases will make it easier to follow the effects of
the next commits.

hnrklssn committed:
[IDE] Support combining --print-module-imports w/ --module-print-hidden

This prints all imports instead of just exported imports, for testing
purposes.

hnrklssn committed:
[ClangImporter] Deduplicate imports copied from submodule to TLM

The imported top-level module inherits the imports of all its
(transitive) submodules. Since multiple submodules can import the same
modules these need to be deduplicated to avoid redundant work.

hnrklssn committed:
[ClangImporter] Only copy submodule imports for top-level modules

Since only top-level Swift modules ever contain code, setting imports
for submodule wrappers has no effect.

hnrklssn commented:
https://github.com/swiftlang/llvm-project/pull/11022
@swift-ci please smoke test

hnrklssn committed:
[ImportResolution] Don't deduplicate scoped imports

This fixes a few tests that were failing because scoped imports were
counted as having imported the module already, leading to later imports
of the same module being skipped. Since the scoped imports don't import
the full module, this would result in syntactically imported symbols not
being found by the compiler.

hnrklssn committed:
[ImportResolution] Override module imports for macros in namespaces

Contents in namespaces are imported to the bridging header module
`__ObjC`. When macros attached to a decl imported from a namespace are
expanded, they also end up in `__ObjC`. This prevents them from
inheriting imports from the module they originated in. This patch
implements a workaround by adding an alternative constructor to
`ImportResolver` that takes an explicit origin module as input,
overriding which module counts as the "parent" module.

hnrklssn commented:
https://github.com/swiftlang/llvm-project/pull/11022
@swift-ci please smoke test

hnrklssn committed:
[ClangImporter] Fix CxxStdlib submodule names in swift interface

A previous commit resulted in the swift interface for CxxStdlib
containing an `import CxxStdlib` for each submodule. This restores the
submodule names to `CxxStdlib.vector` etc.

hnrklssn commented:
https://github.com/swiftlang/llvm-project/pull/11022
@swift-ci please smoke test

hnrklssn commented:
Backed out import printing changes to unblock this while we're deciding what we want to do there. Tracked as rdar://160514023.

hnrklssn commented:
https://github.com/swiftlang/llvm-project/pull/11022
@swift-ci please smoke test linux platform

hnrklssn commented:
https://github.com/swiftlang/llvm-project/pull/11022
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
handle libc++ v17-19

libc++ v17-19 was split into multiple top-level modules, while versions
earlier and later had one TLM with submodules.

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test linux

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
[ImportResolution] Gracefully handle importing broken clang module

An assert checking the invariant of the module passed to
alreadyImportedTLM being a top-level module would dereference a null
pointer in the case where the clang module contained syntax errors,
since findUnderlyingClangModule would return null. Instead call the
bespoke isSubmodule function that already performs this null check.

This fixes the lldb test
lldb/test/API/lang/swift/clangimporter/expr_import/TestSwiftExprImport.py.

hnrklssn commented:
@swift-ci please smoke test

AnthonyLatsis commented:
This seems to have broken rebranch: https://ci.swift.org/view/Swift%20rebranch/job/oss-swift-rebranch-incremental-RA-macos/7016. Confirming in https://github.com/swiftlang/swift/pull/81190.
